### DISCUSSION TITLE
What are the possible reasons for deadlock of `tokio::sync::Mutex`?

### DISCUSSION THREAD

**Discussion by rise0chen** [tokio-rs/tokio#5176] [Q&A] [answered]
What are the possible reasons for deadlock of `tokio::sync::Mutex`?

**Comment by Darksonn** [comment, marked as answer]
The same reasons as any other mutex in any programming language. Any textbook on concurrency will talk about this.

The short story is that if you need to lock two mutexes at the same time, you must make sure that you always lock them in the same order. If you don't, then this is very likely to result in a deadlock.
